Catch images and captions send to editor from the media library

goldenapples opened the pull request:
Adds a filter on `media_send_to_editor` to catch images selected through the media library; builds out the callback function to handle links and captions as well.

goldenapples commented:
@danielbachhuber can you give this a review?

danielbachhuber commented on inc/class-img-shortcode-integrations.php:
```diff
@@ -0,0 +1,51 @@
+			}
+		}
+
+		$shortcode .= '/]';
```
Maybe we should make this shortcode generation method a method on the `Img_Shortcode` class?

goldenapples replied:
Sounds good, its a simple method and I don't have anything else to go in an "integrations" class.

053eda6

danielbachhuber commented on inc/class-img-shortcode.php:
```diff
@@ -123,7 +123,7 @@ public static function callback( $attr ) {
 			'caption'    => '',
-			'align'      => 'none',
+			'align'      => 'alignnone',
 			'linkto'     => '',
```
for `linkto`, we should respect the value of `image_default_link_type` option.

goldenapples replied:
Yes, but not here. That default should be applied when adding the image in the first place, to follow the expected behavior in the Media Library UI. 

I don't see a way to set a default value for a field when registering it with `shortcode_ui_register_for_shortcode()` in Shortcake. Am I missing that, or is that possible to do currently?

If not, I could track down the code and make a pull request for Shortcake to add that capability.

danielbachhuber replied:
> I don't see a way to set a default value for a field when registering it with shortcode_ui_register_for_shortcode() in Shortcake. Am I missing that, or is that possible to do currently?

I guess we don't have it. I've thought we did. `default_value` would be a good argument to pass.

danielbachhuber commented on inc/class-img-shortcode.php:
```diff
@@ -173,7 +173,105 @@ public static function callback( $attr ) {
+	 *
+	 */
+	private static function linkify( $img_tag, $attributes ) {
+		$html = apply_filters( 'img_shortcode_link_output', '', $img_tag, $attributes );
```
Rather than follow the short circuit pattern, I'd prefer to filter at the end of the method. The developer can always replace the contents of the return value with whatever they want to short circuit. However, filtering at the end permits minor modification of existing markup.

goldenapples replied:
I agree. Looking at usage elsewhere, it seems apparent that the short-circuit pattern in a couple shortcodes in core is an abandoned design pattern, and a filter that gets some input is much more useful than one which doesn't.

I moved all of the filters to the `callback()` method though; to keep them all on one place and to handle conditions where you would want to filter output after a certain step, whether or not the `linkify` or `captionify` functions are called.

184a252

danielbachhuber commented:
Glad `media_send_to_editor` exists :) Otherwise, this would likely be a massive hack.

Looking pretty good so far. Some minor notes.

goldenapples commented:
Responded to some code review comments. This can be evaluated in the theme with this branch: https://github.com/fusioneng/fusion-theme/pull/2669

danielbachhuber commented on image-shortcake.php:
```diff
@@ -25,6 +25,7 @@ public static function get_instance() {
 
 			self::$instance = new Image_Shortcake;
 			self::$instance->register_shortcode();
+			self::$instance->attach_filters();
```
Can we call this `setup_filters()` for consistency with our other plugins?

danielbachhuber commented on inc/class-img-shortcode.php:
```diff
@@ -111,7 +111,13 @@ public static function reversal( $content ) {
+	 * - `img_shortcode_attrs`          Initial shortcode attributes
+	 * - `img_shortcode_img_output`     Output of the <img> tag, before wrapping in link and caption
+	 * - `img_shortcode_link_output`    The <img> tag, possibly wrapped in a link, before wrapping in caption
+	 * - `img_shortcode_caption_output` Final shortcode output: caption, link, and image
```
Should we adhere to WordPress' PHPdoc standards for inline filters?

danielbachhuber commented on inc/class-img-shortcode.php:
```diff
@@ -173,7 +163,139 @@ public static function callback( $attr ) {
+			}
+
+			$image_html = self::captionify( $image_html, $attr );
+		}
```
Shouldn't we return early if there isn't a caption to apply? I wouldn't expect my markup to go through the caption filter if I didn't have a caption on the post.

goldenapples replied:
Would renaming the filters make it make more sense?

I imagine there would be cases where you'd want to filter after the caption was added, so that markup could be added to the very beginning or the very end - whether or not there was a caption applied.

Like how at https://github.com/fusioneng/fusion-theme/blob/eaaa259a30b2f6bc3bdeb9499171076e70e78152/inc/frontend/class-inline-image.php#L58-79 I attached a filter to `img_shortcode_link_output` because I wanted the share links to go outside the link if one existed, or just after the image if there was no link.

goldenapples replied:
I renamed the filters in 594812c. Is this more clear?

danielbachhuber replied:
> I renamed the filters in 594812c. Is this more clear?

Yep, it's more explicit.

danielbachhuber commented:
Some small comments.

goldenapples commented:
Addressed some comments. I think that giving the filters clearer names makes them more readable. I can add filters inside the `linkfy` and `captionify` functions if it'd be useful, but I can't immediately think of a use case for them, so I left the filters in the place they were. Review?
